### PR TITLE
Separate ELK ES client config from custom index ES client config

### PR DESCRIPTION
**Pull request by ethanewing**
Ready to merge

### CODE REVIEW

**Comment by nickmarden** [comment]
@ethanewing looks like you need to reformat the `secrets.yml` file used by the test infrastructure, to take advantage of the new expectations around ES endpoint information.

Take a gander through the `.circleci/config.yml` file to see how that configuration file gets set up during the testing process.

**Comment by nickmarden** [comment]
So close!

```
=========== Lines missing coverage: ===========
/home/circleci/search-gov/lib/es.rb, 31
Coverage (99.99%) is below the expected minimum coverage (100.00%).
```

**Comment by ethanewing** [comment]
(This is all in reference to the following change: [https://github.com/GSA/search-gov/pull/74/commits/da6ed84c5840720ac982051fc608d0fa3c192628](url)
Just for clarity, I thought I'd explain how the `ES` module and its submodules are working, since the significance/correctness of the changes I've made might not be readily apparent.

The issue I was encountering prior to this change was that anytime that, say, `ES::ELK.client_reader` was called, then the class variable `@@client_reader` would be initialized to correspond to that particular analytics node for all submodules of `ES` -- future calls to `ES::CustomIndices.client_reader` or `ES::ELK.client_reader` would always use that particular node, which isn't the behavior we want if we want different parts of the program to interact with analytics nodes and index nodes.  This issue is rectified when the class variables in the `ES` module are changed to instant variables.  I'll outline why I think this change works through the following steps:
1. I call `ES::ELK.client_reader`
2. The program goes into the `ES::ELK` submodule, which has .client_reader extended from `ES`
3. `@client_reader` is then initialized in that call, but it is an instance variable of only `ES::ELK`, because that call was only being made within the scope of `ES::ELK`; even though the function client_reader is defined in the parent module, that function (and thus the instance variable declared and assigned within it) is only ever called within the scope of `ES::ELK`, so this creates an instance variable that only the `ES::ELK` submodule can see -- it's invisible to `ES` and `ES::CustomIndices`.
Basically, if you just imagine that those two functions were written within each of the submodules (in which case the file would not be DRY), how everything is being scoped becomes more intelligible.  `ES`, `ES::ELK`, and `ES::CustomIndices` all have their own instance variables, even if there's repetition in their names.

**Comment by nickmarden** [comment]
@ethanewing some requests:
- Rebase against the current and squash-merge your commits into a single commit
- Rename the PR something useful, such as the title of the original issue ("Separate ELK ES client config from custom index ES client config")
- Remove the 'do-not-merge' label

File: config/secrets.yml.dev [hunk 1]
```diff
@@ -55,9 +55,14 @@ dev_settings: &DEV_SETTINGS
   asis:
     host: http://localhost:8080
   elasticsearch:
-    reader: localhost
```
[the remark this answers is not in the document]
Comment (noremmie) [reply config/secrets.yml.dev]: @nickmarden I agree that this is a minor nit.

I also agree that `elasticsearch:analytics:reader` is more clear than `elasticsearch:reader:analytics`.

I disagree that `analytics:elasticsearch:reader` is more clear than `elasticsearch:analytics:reader`.

[the remark this answers is not in the document]
Comment (nickmarden) [reply config/secrets.yml.dev]: My thinking about the approach of `analytics:...` is that these ES settings are _currently_ the only information associated with our configuration for analytics and custom indices, but in the future they might not be.

Imagine that later we wanted to add some information about, say, the URL at which we publicly expose the analytics Kibana 5 dashboard. In this case, `kibana5.search.usa.gov` might be a bad name if the Kibana 5 interface for the ASIS/i14y/etc. cluster is a different one than the Kibana5 interface for the analytics (logstash) cluster. So we might want to associate a `dashboard_url` attribute with the analytics portion of our system.

I was thinking that in that case it might be a more obvious organization of attributes to say

```
analytics:
  dashboard_url: https://analytics.search.gov
  elasticsearch:
    reader: ...
    writers:
      - ...
```

than it would be to say

```
elasticsearch:
  analytics:
    dashboard_url: https://analytics.search.gov
    reader: ...
    writers:
      - ...
```

Because the latter implies that the `dashboard_url` is an Elasticsearch setting, whereas the former implies that is a characteristic of our analytics subsystem.

Yes, this is a nit. But since we're reorganizing the app and the corresponding Chef recipes anyway, I thought this was the best time to bring it up.

[the remark this answers is not in the document]
Comment (noremmie) [reply config/secrets.yml.dev]: I think we're on the same page then, @nickmarden.  My thinking was that it would only make sense to use `analytics:elasticsearch:reader` if there was something else other than `elasticsearch` that would live underneath `analytics`, and I couldn't think of such a thing that we would ever add there. But you've thought of such a thing: `dashboard_url`.

**Comment by nickmarden** [comment]
OK @ethanewing please make the above-discussed change.

The settings in the `config.yml` file, and in the corresponding `Rails.secrets` calls, should be:
```
analytics:elasticsearch:reader
analytics:elasticsearch:writers
custom_indices:elasticsearch:reader
custom_indices:elasticsearch:writers
```

The corresponding ES calls will now probably look something like

```
Rails.application.secrets.send(self.client_config).elasticsearch['writers']
```

etc. See [Object#send](https://ruby-doc.org/core-2.5.1/Object.html#method-i-send) for an explanation of what I mean, and remember that the thing that you `#send` needs to be a symbol, not a string.

**Comment by nickmarden** [comment]
LGTM 👍 

@noremmie let's coordinate the merge of this with a corresponding recipe change that I suppose I'll have to write :-)

**Comment by nickmarden** [comment]
This is ready for testing on demo3

**Comment by noremmie** [comment]
Understood, @nickmarden. This situation is a little weird because normally the developer would list out all the acceptance criteria for the story, but in this case that would probably be a huge amount of instructions. I'm assuming you're doing the acceptance testing, but I can spend some time tomorrow helping to kick the tires on demo3.

**Comment by noremmie** [comment]
@ethanewing and @nickmarden sorry for being late to the PR party. As penance for that, I've pushed commit 4e021e8 that makes the changes I would have otherwise suggested in comments. Please let me know if either of you object to these changes, although I suspect you will 💖 them. These are just spec changes that won't invalidate any testing that has already been done on demo3.

I'll do some of my own testing on demo3 today too.

**Comment by nickmarden** [comment]
Thanks for the spec fixes @noremmie - good catches.

@ethanewing plz inspect your commits for errant `p foo` and `byebug` commands before pushing to master ;-)

File: lib/es.rb [hunk 2]
```diff
@@ -3,23 +3,41 @@
 module ES
   INDEX_PREFIX = "#{Rails.env}-usasearch".freeze
 
-  def self.client_reader
-    @@client_reader ||= Elasticsearch::Client.new(log: Rails.env == 'development', host: reader_host)
```
Comment (noremmie) [inline lib/es.rb]: I was concerned that the change from `@@client_reader` to `@client_reader` would cause these clients to be re-created much more frequently than they were before (and therefore new ES connections to be created much more frequently than before). But I added some logging on demo3 that would show me whenever these new clients are created. I saw that connections were still being re-used, so I'm no longer worried.

In retrospect this makes sense because these are actual module instance variables and not object instance variables, so they ultimately belong to the classes this module gets mixed into, not instances. Once a particular class creates its `@client_reader`, it holds onto it for the rest of its life 👍

**Comment by noremmie** [comment]
My testing is looking good so far. The only remaining items are to test the search-gov custom indices, and I've switched from demo3 to demo1 for that because the custom indices there have more data. I'll be able to wrap this up early tomorrow morning barring disaster.

**Comment by noremmie** [comment]
I completed my testing on demo1, and the search-gov app's interaction with the custom indices looks good to me. I'll look into rolling this out today keeping in mind that there's a corresponding cookbooks update that needs to accompany the release.

**Comment by noremmie** [comment]
FYI @ethanewing I just deployed these changes, and I don't see any sign of trouble from production. Nice work!

One thing I neglected to mention about testing earlier is that I couldn't test the `WatcherObserver` updates in demo environments because the Elasticsearch licenses in those demo environments don't support watchers. But I just tested one in production, and it worked great 👍

**Comment by ethanewing** [comment]
Awesome! I'm glad everything's working as it should.
